feat(core): Flush chunks on hourly boundaries.

broneill opened the pull request:
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [ ] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :**
Chunks are flushed by a background task.

**New behavior :**
Chunks are flushed when ingestion time crosses an hourly boundary. The existing background task remains unchanged for now, but perhaps it can run less frequently.

velvia commented on core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala:
```diff
@@ -429,6 +440,19 @@ class TimeSeriesShard(val ref: DatasetRef,
           }
         } else {
+          val oldTimestamp = groupTimestamps(group)
+          val newTimestamp = Math.max(oldTimestamp, ingestionTime) // monotonic clock
```
This check should not happen within `onNext`, as the ingestion time does not change for each sample.  This makes the check much more expensive than it should be ... just check it after each call to ingest().

broneill replied:
Ingestion time can vary by group, and the monotonic adjustment shouldn't span groups.

velvia commented on core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala:
```diff
@@ -429,6 +440,19 @@ class TimeSeriesShard(val ref: DatasetRef,
+              case None => // none provided, so don't flush
+              case Some(sched) => {
+                // Flush out the group first such that the record is for a new hour.
+                createFlushTask(prepareFlushGroup(group)).executeOn(sched)
```
This logic does things on a different thread, and should not be here - it pollutes the IngestConsumer.

broneill replied:
It looks like the original task was scheduled on the same thread. I'll check again.

velvia commented on core/src/main/scala/filodb.core/memstore/TimeSeriesMemStore.scala:
```diff
@@ -123,17 +122,12 @@ extends MemStore with StrictLogging {
     val shard = getShardE(dataset, shardNum)
     combinedStream.map {
-                    case d: SomeData =>       shard.ingest(d)
+                    case d: SomeData =>       shard.ingest(d.records, d.offset, Some(flushSched))
```
I don't think ingest() should involve itself in triggering the flush at all. 

I think a much cleaner separation would be like this:
  1) Remove the `FlushCommand`... that won't be necessary any more
  2) Code could be like this:

     case d: SomeData => shard.ingest(d)
                                           if (shard.checkTimeToFlush()) {
                                              Some(shard.prepareFlushGroup(group))
                                           } else None

broneill replied:
Yes. I left the FlushCommand in for now just to simplify the review.

broneill replied:
The reason why ingest triggers the flush is that it cannot absorb any more data until the flush has been prepared. If some stateful check must be invoked on the shard after ingest has been called, it's too late.

velvia replied:
This can be accomplished using the example I gave above.  It's not too late, as long as it happens before the next data is ingested.

broneill replied:
Another reason why I decided to have the shard trigger the flush is that it does so as it's iterating over the record container. If TimeSeriesMemStore needed to trigger the flush, then the record container would need to be iterated over twice (or more).

velvia replied:
We don't need to iterate over the record container to trigger the flush - there is only one ingestion time per recordContainer.  The trigger can be done before or after.

broneill replied:
True dat.

broneill replied:
...however, there's the monotonic clock check, which is adjusted per group. A monotonic clock per record container requires switching to 1 flush group, which I'm hoping to avoid.

velvia replied:
> there's the monotonic clock check, which is adjusted per group

How does this work again?  The per group check?

broneill replied:
The per group check is performed as sampled are extracted from the record container. which is heterogeneous.

velvia commented on core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala:
```diff
@@ -127,6 +127,11 @@ object TimeSeriesShard {
+  // Flush groups when ingestion time is observed to cross an hourly boundary. This simplifies
+  // disaster recovery -- chunks can be copied without concern that they may overlap in time.
+  // TODO make configurable if necessary
+  val flushBoundaryMillis = 1.hour.toMillis
```
See `StoreConfig`... there is a configuration already for the flush interval, we should just use that.

broneill replied:
ok

broneill commented:
I've made changes such that the flushing is distributed across the hour (whatever flush interval is) by the number of flush groups. Note that the flush stream is completely gone now.

I modified the tests that were failing, and I removed one that's now invalid (it tested the removed flush stream code). The tests might still need a bit more work, but they do demonstrate that flushing by ingestion time is working as expected.

velvia commented:
@broneill  just to let you know - I haven't forgotten about this, just looked through all of the new changes.  I need to absorb the changes and the design a bit more.... specifically, why the change from initiating the flush tasks from within the ingest method itself from the original method.   I'm afraid that is still a very big change and would like to keep the original processing the way it is if possible.  But let me think about this some more.

broneill commented on core/src/main/scala/filodb.core/memstore/TimeSeriesMemStore.scala:
```diff
@@ -101,49 +100,24 @@ extends MemStore with StrictLogging {
-                  .completedL
-                  .doOnCancel(cancelTask)
-                  .runAsync(shard.ingestSched)
+    stream.map { case d: SomeData => shard.ingest(d.records, d.offset, Some(flushSched)) }
```
I think this is vestigial and can be removed.

broneill replied:
Actually, it's still required. Some methods like TimeSeriesMemstore.ingest don't have access to a flushSched.

broneill commented on core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala:
```diff
@@ -462,6 +479,21 @@ class TimeSeriesShard(val ref: DatasetRef,
+          if (newTimestamp > oldTimestamp && oldTimestamp != Long.MinValue) {
+            val tsAdjust = flushOffsetMillis * group
+            if ((oldTimestamp - tsAdjust) / flushBoundaryMillis != (newTimestamp - tsAdjust) / flushBoundaryMillis) {
+              flushSched match {
```
Sounds right to me.

broneill replied:
Actually, I don't think these extra calculations can be omitted. I need to make a precise timestamp check before rounding down.

broneill commented on core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala:
```diff
@@ -462,6 +479,21 @@ class TimeSeriesShard(val ref: DatasetRef,
+                case None => // none provided, so don't flush
+                case Some(sched) => {
+                  // Flush out the group first such that the record is for a new hour.
+                  createFlushTask(prepareFlushGroup(group)).executeOn(sched)
```
How is this done? I'm not sure what the correct pattern to apply is.

velvia replied:
Ah, I see that Vish called out the same issue here.

velvia commented:
Just chatted with @broneill offline, but adding it here for the record.
He's going to explore an approach where we can check the flush trigger outside of `IngestConsumer`, that way it won't need to be done on a per-sample basis, and allow us to keep the previous `mapAsync` approach of initiating the flush tasks, which keeps the ability to limit the number of simultaneous flush tasks going at the same time (for write load management).
